File: jobs/pull_favicons.js
var MongoClient = require('mongodb').MongoClient
var config = require('../config');
var request = require('request');
var http = require('http');
var fs = require('fs');
var express = require('express');


var cheerio = require('cheerio');

var environment = process.env.NODE_ENV || "test";

console.log(environment);

if (environment == "production"){
	mongo_user = process.env.MONGO_USER;
	mongo_pw = process.env.MONGO_PW;
	mongo_db = process.env.MONGO_DB;

} else {
	mongo_user = config.mongodb.user_name;
	mongo_pw = config.mongodb.user_pass;
	mongo_db = config.mongodb.db_name;
}


var mongo_link = "mongodb://"+mongo_user+":"
	+mongo_pw+"@ds151070.mlab.com:51070/startpage-test";

console.log("mongoDB link: " +mongo_link)	

MongoClient.connect(mongo_link, (err, client)=>{
	if (err) return console.log(err)
	db = client.db(mongo_db);
	
	console.log("Connected to MongoDB");

	//scrape each site in database collection for favicons
	//get all bookmarks
	var bm_list;
	db.collection("bookmarks").find().toArray(function(err, result){
		if (err) return console.log(err)
		console.log("Pulled bookmarks");
		bm_list = result;

		console.log(bm_list);

		//for each bookmark
		for (var x =0; x<bm_list.length; x++){
			//get url
			var query_url = bm_list[x].bookmark_url
			var favicon_name = bm_list[x].bookmark_name + ".ico"
			var request = http.get("http://i3.ytimg.com/vi/J---aiyznGQ/mqdefault.jpg", function(response) {
			  response.pipe(favicon_name);
			});
			query_url = 'http://www.google.com/s2/favicons?domain=' + query_url;
			console.log(query_url);

		};
	});

	client.close();
	
});

File: python_jobs/pull_favicons.py
import urllib.request
import pymongo
import json
import os
import boto3

from pymongo import MongoClient


try:
	if(os.environ['NODE_ENV']=='production'):
		mongo_user = os.environ['MONGO_USER']
		mongo_pw = os.environ['MONGO_PW']
		mongo_db = os.environ['MONGO_DB']
		access_key = os.environ['AWS_ACCESS_KEY']
		secret_key= os.environ['AWS_SECRET_KEY']
		bucket_name = os.environ['S3_BUCKET_NAME']
		aws_region = os.environ['S3_REGION']
	else:
		config_path=(os.path.join(os.path.dirname(__file__), 'config_py.json'))
		configs = json.load(open(config_path))
		mongo_user = configs['mongodb_user_name']
		mongo_pw = configs['mongodb_user_pass']
		mongo_db = configs['mongodb_db_name']
		access_key = configs['aws_access_key']
		secret_key = configs['aws_secret_key']
		bucket_name = configs['s3_bucket_name']
		aws_region = configs['s3_region']
except: 
	config_path=(os.path.join(os.path.dirname(__file__), 'config_py.json'))
	configs = json.load(open(config_path))
	mongo_user = configs['mongodb_user_name']
	mongo_pw = configs['mongodb_user_pass']
	mongo_db = configs['mongodb_db_name']
	access_key = configs['aws_access_key']
	secret_key = configs['aws_secret_key']
	bucket_name = configs['s3_bucket_name']
	aws_region = configs['s3_region']
	
mongo_link = "mongodb://"+mongo_user+":"+mongo_pw+"@ds151070.mlab.com:51070/startpage-test"
client = MongoClient(mongo_link)
print("mongodb link: "+mongo_link)

db = client[mongo_db]
bookmarks = db['bookmarks']
print(bookmarks)

for bm in bookmarks.find():
	#get urlpymongo

	print(bm['bookmark_url'])
	fav_url = bm['bookmark_url']
	fav_name = bm['bookmark_name']+'.ico'
	#create favicon link
	fav_url = 'http://www.google.com/s2/favicons?domain=' + fav_url
	#download favicon
	urllib.request.urlretrieve(fav_url, fav_name)
	#upload to s3
	s3_client = boto3.client(
		's3',
		aws_access_key_id = access_key,
		aws_secret_access_key = secret_key)

	to_s3_icon = open(fav_name, 'rb')
	s3_client.put_object(Bucket=bucket_name,Key=("bm/"+fav_name), Body = to_s3_icon)
	to_s3_icon.close()
	print(fav_name + "Put out to S3")
	#delete local favicon
	os.remove(fav_name)
	#assign key to db entry for bookmark
	bookmarks.update_one({'bookmark_name': bm['bookmark_name']},
		{'$set': {'bookmark_favicon': ("bm/"+fav_name)}})
	print("mongodb updated")